bson/value_reader_test.go

```diff
@@ -176,16 +175,16 @@ func TestValueReader(t *testing.T) {
 			}
 
 			// invalid length
-			vr.r = bufio.NewReader(bytes.NewReader([]byte{0x00, 0x00}))
+			vr.src = &bufferedValueReader{buf: []byte{0x00, 0x00}}
 			_, err := vr.ReadDocument()
-			if !errors.Is(err, io.ErrUnexpectedEOF) {
+			if !errors.Is(err, io.EOF) {
 				t.Errorf("Expected io.ErrUnexpectedEOF with document length too small. got %v; want %v", err, io.EOF)
```

Comment:
`Read*` methods should return `io.ErrUnexpectedEOF` if the data ends before the requested value can be read. From the `io.EOF` docs:
> ... Functions should return EOF only to signal a graceful end of input. If the EOF occurs unexpectedly in a structured data stream, the appropriate error is either [ErrUnexpectedEOF] or some other error giving more detail.

Reply:
In the buffered path, EOF is returned from `peek`, not from attempting a read. Returning `io.ErrUnexectedEOF` doesn't' make sense in that context. Additionally, using bufio discard would just return `io.EOF` if a user tries to discard more data than was on the buffer.

Reply:
Sorry, I was talking about the structured `Read*` methods, like `ReadDocument` or `ReadBinary`. I see now that many of the existing tests expect `io.EOF` from those methods ([example](https://github.com/mongodb/mongo-go-driver/blob/9d878b710ac8d254d51eaa19bca8c87e03ebe2b1/bson/value_reader_test.go#L773)), but it's inconsistent between the different methods. We should either preserve the existing error behavior or standardize on returning `io.UnexpectedEOF`. The unstructured stream readers (`streamingValueReader` and `bufferedValueReader`) should return `io.EOF`.

Reply:
@matthewdale I've created GODRIVER-3613 to investigate this further, per our conversation.